### PR TITLE
Improve Font Metadata APIs

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/Fonts/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

There's a fair amount of breaking refactoring going on here but I deem this necessary.

Touches #86 
- Refactors the `**Instance` classes surrounding metadata to make usage more obvious using `**Metrics` as a new suffix. 
- Deduped the `Font`/`FontMetrics` properties.
- Renamed existing `GlyphMetric` struct to `GlyphBounds`.
- Updated XML, documentation to reflect changes.
- Exposed `CodePoint` in all public APIs to improve sanitation and added more unit tests.

There's still a few things missing from the `GlyphMetrics` - e.g Width, + vertical properties. But I wanted to discuss them first.
_Edit : I added support for vhea and vmtx tables and exposed the equivalent properties._

Here's some unit tests demonstrating the new API.

`FontMetrics`

```c#
[Fact]
public void FontMetricsMatchesReference()
{
    // Compared to FontForge metrics
    var collection = new FontCollection();
    FontFamily family = collection.Install(TestFonts.OpenSansFile);
    Font font = family.CreateFont(12);

    Assert.Equal(2048, font.FontMetrics.UnitsPerEm);
    Assert.Equal(font.FontMetrics.UnitsPerEm * 72F, font.FontMetrics.ScaleFactor);
    Assert.Equal(2189, font.FontMetrics.Ascender);
    Assert.Equal(-600, font.FontMetrics.Descender);
    Assert.Equal(0, font.FontMetrics.LineGap);
    Assert.Equal(2789, font.FontMetrics.LineHeight);
    Assert.Equal(2476, font.FontMetrics.AdvanceWidthMax);
    Assert.Equal(font.FontMetrics.LineHeight, font.FontMetrics.AdvanceHeightMax);
}
``` 

`GlyphMetrics`

```c#
[Fact]
public void GlyphMetricsMatchesReference()
{
    // Compared to FontForge metrics
    var collection = new FontCollection();
    FontFamily family = collection.Install(TestFonts.OpenSansFile);
    Font font = family.CreateFont(12);

    var codePoint = new CodePoint('A');
    GlyphMetrics glyphMetrics = font.FontMetrics.GetGlyphMetrics(codePoint);
    GlyphMetrics glyphMetrics1 = font.GetGlyph(codePoint).GlyphMetrics;

    Assert.Equal(glyphMetrics, glyphMetrics1);

    Assert.Equal(codePoint, glyphMetrics.Codepoint);
    Assert.Equal(font.FontMetrics.UnitsPerEm, glyphMetrics.UnitsPerEm);
    Assert.Equal(glyphMetrics.UnitsPerEm * 72F, glyphMetrics.ScaleFactor);
    Assert.Equal(1296, glyphMetrics.AdvanceWidth);
    Assert.Equal(2789, glyphMetrics.AdvanceHeight);
    Assert.Equal(1296, glyphMetrics.Width);
    Assert.Equal(1468, glyphMetrics.Height);
    Assert.Equal(0, glyphMetrics.LeftSideBearing);
    Assert.Equal(0, glyphMetrics.TopSideBearing);
    Assert.Equal(GlyphType.Standard, glyphMetrics.GlyphType);
}
```

Using `CodePoint` everywhere adds a little noise when dealing with `ReadonlySpan<char>` but people really shouldn't be iterating that anyway. I can add an extension method to that  type allowing us to enumerate `CodePoint`s instead.

### CODE REVIEW

File: src/SixLabors.Fonts/Font.cs [hunk 1]
```diff
@@ -73,99 +74,65 @@ public Font(Font prototype, float size)
...
         /// <summary>
-        /// Gets the line height.
+        /// Gets the font metrics.
         /// </summary>
-        public int LineHeight => this.Instance.LineHeight;
+        public IFontMetrics FontMetrics => this.metrics.Value ?? throw new FontException("Font instance not found.");
 
         /// <summary>
-        /// Gets the font instance.
+        /// Gets the requested size.
```
Comment (bijington) [inline src/SixLabors.Fonts/Font.cs:120]: Requested 'style'?

File: src/SixLabors.Fonts/Glyph.cs [hunk 2]
```diff
@@ -12,31 +12,28 @@ namespace SixLabors.Fonts
     {
         private readonly float pointSize;
 
-        internal Glyph(GlyphInstance instance, float pointSize)
+        internal Glyph(GlyphMetrics instance, float pointSize)
```
Comment (bijington) [inline src/SixLabors.Fonts/Glyph.cs:15]: minor but `metrics` is probably a clearer name than `instance`

File: src/SixLabors.Fonts/GlyphMetrics.cs [hunk 3]
```diff
@@ -5,92 +5,105 @@
...
 
+// TODO: This should support both horizontal and vertical metrics via vmtx table.
 namespace SixLabors.Fonts
 {
     /// <summary>
-    /// A glyph from a particular font face.
+    /// Represents a glyph metric from a particular font face.
     /// </summary>
-    public partial class GlyphInstance
+    public class GlyphMetrics
```
Comment (bijington) [inline src/SixLabors.Fonts/GlyphMetrics.cs:16]: I hate to be really picky here but if it does represent a single glyph metric should the class be `GlyphMetric`?

Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/GlyphMetrics.cs:16]: Yesn’t…
It’s a single glyph but multiple metric measurements. The generally used term for similar types in this field is the plural.

File: src/SixLabors.Fonts/Glyph.cs (hunk 2 above)
Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/Glyph.cs:15]: Yep, missed that!

File: src/SixLabors.Fonts/Font.cs (hunk 1 above)
Comment (JimBobSquarePants) [reply src/SixLabors.Fonts/Font.cs:120]: Yup!

File: src/SixLabors.Fonts/GlyphMetrics.cs (hunk 3 above)
Comment (bijington) [reply src/SixLabors.Fonts/GlyphMetrics.cs:16]: Ah that makes sense thanks

**Pull request merged**